Fix invalid literals in generated rst docs

Jayman2000 opened the pull request:
Before this change, API docs that look like this:

	…adding [param character]s to the right of the string.

would turn into rst files that look like this:

	…adding ``character``s to the right of the string.

That reStructuredText is invalid and causes warnings when the docs repo is built.

<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.
-->

YuriSizov commented:
Would be worth adding a comment in there to explain why we do this. But also, doesn't this always add an extra empty space after the tag? And what about the linked references, does that not break them?

Btw, what doc is this quote from? It should be rephrased, because we shouldn't use code identifiers in text like that.

Jayman2000 committed:
Fix invalid literals in generated rst docs

Before this change, API docs that look like this:

	…adding [param character]s to the right of the string.

would turn into rst files that look like this:

	…adding ``character``s to the right of the string.

That reStructuredText is invalid and causes warnings when the docs repo
is built.

Jayman2000 commented:
> Would be worth adding a comment in there to explain why we do this.
> […]
> This should also be fixed for the overall else case.

OK, so it turns out that there actually is [something that’s supposed to fix this for the entire `format_text_block` function](https://github.com/godotengine/godot/blob/79b21e96ad7a154e6c64a88150c98bd4dea52e91/doc/tools/make_rst.py#L1582), and it even includes a comment that explains why it does what it does:

```python
        # Properly escape things like `[Node]s`
        if escape_pre and pre_text and pre_text[-1] not in MARKUP_ALLOWED_PRECEDENT:
            pre_text += "\ "
        if escape_post and post_text and post_text[0] not in MARKUP_ALLOWED_SUBSEQUENT:
            post_text = "\ " + post_text
```

In fact, the first hunk in my commit was completely pointless. My first hunk was dealing with this if statement:

```python
        if tag_text in state.classes:
            if tag_text == state.current_class:
                # Don't create a link to the same class, format it as inline code.
                tag_text = f"``{tag_text}``"
            else:
                tag_text = make_type(tag_text, state)
            escape_pre = True
            escape_post = True
```

Enabling `escape_pre` and `escape_post` at the end makes sure that escaped spaces are added if they’re needed.

In reality, the bug wasn’t that escaped spaces weren’t being added. The bug is the fact that they weren’t being added __when a `[param foo]` tag is used__. I just pushed a commit that fixes this bug properly.

---

> But also, doesn't this always add an extra empty space after the tag?

It won’t add an an extra space. According to [the reStructuredText Specification](https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#escaping-mechanism):

> Escaped whitespace characters are removed from the output document together with the escaping backslash. This allows for [character-level inline markup](https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#character-level-inline-markup-1).

---

> And what about the linked references, does that not break them?

I don’t think so. Could you give an example of some markup that might break?

---

> Btw, what doc is this quote from? It should be rephrased, because we shouldn't use code identifiers in text like that.

Here’s a list of everything that I’m aware of that does something like this: ``` ``foo``s ```

- <https://docs.godotengine.org/en/latest/classes/class_astar2d.html#class-astar2d-method-add-point>
- <https://docs.godotengine.org/en/latest/classes/class_astar3d.html#class-astar3d-method-add-point>
- <https://docs.godotengine.org/en/latest/classes/class_string.html#class-string-method-lpad>
- <https://docs.godotengine.org/en/latest/classes/class_string.html#class-string-method-rpad>
- <https://docs.godotengine.org/en/latest/classes/class_tree.html#class-tree-method-create-item>

YuriSizov reviewed (approved):
Makes sense now, thanks.

YuriSizov commented:
Thanks!
